### ISSUE TITLE
schema.core/defn false positive on arity 

### ISSUE THREAD

**Issue by ccann** [borkdude/clj-kondo#106]
Hi!

I noticed a false positive when defining a function with `schema.core/defn` it seems to consider, e.g.

```clojure
(s/defn verify-signature :- Bool
  [message :- Str base64-encoded-signature :- Str]
  ,,,
```

as a function that takes 6 args not 2. I got the corresponding error:
```
<redacted>:79:13: error: wrong number of args (2) passed to <redacted>/verify-signature
```

(this report presupposes you intend to support Schema)

**Comment by borkdude** [comment]
Thanks for the report!

**Comment by borkdude** [comment]
@ccann I'm working on a fix for this. I want to exclude the false positive. Supporting these Schema `defn` macros is something I might want to do if more users use these. Are you using these a lot?

**Comment by ccann** [comment]
Very cool! Yes, our team uses them quite a bit. 

macOS (Mojave, High Sierra) for development and the clojure-alpine docker image for builds (considering adding this to our linting chain at build time, depending on if we can ignore certain rules or namespaces or functions since false positives are difficult to completely eliminate in any linter)

**Comment by borkdude** [comment]
@ccann You can disable certain rules and exclude files from the output. See https://github.com/borkdude/clj-kondo/blob/master/doc/config.md. Let me know if this config works for you, it hasn't seen a lot of usage yet.

**Comment by borkdude** [comment]
I think I can make this work:

![Screenshot 2019-04-24 00 01 27](https://user-images.githubusercontent.com/284934/56618973-57d5dc80-6624-11e9-89a3-69c0abd6c1f4.png)

**Comment by borkdude** [comment]
@ccann Do you by any chance have some public code I could lint to test this more?

**Comment by ccann** [comment]
Unfortunately I don't have any public code I can share. I would if I could!

I did run your binary on some of our older projects and found some other cases to consider:

```clojure
(s/defn foo
  [a :- s/Int b :- s/Bool])

;; expects 6 args intead of 2 (as already reported)
(foo 1 true)

;; can't parse <ns>, clojure.lang.PersistentVector cannot be cast to clojure.lang.Named
(s/defn bar
  [{:keys [foos :- [Baz]]}])

;; can't parse <ns>, No matching clause: :display
(s/defn ->color
  [template & [filters :- {:display s/Bool}]])

;; can't parse <ns>, Don't know how to create ISeq from: clojure.lang.Symbol
(s/defn foo :- (s/maybe s/Int)
  [a :- s/Int b :- s/Bool])
```

Funny enough, someone had written:
```clojure
(s/defn count-matches :- [{:palette :- cu/PalleteSchema :count :- s/Num}]
  [pcols :- #{s/Str} palettes :- [cu/PalleteSchema]])
```

which has improper syntax for the return type (should omit the `:-` symbols inside a map) which caused the linter to report `can't parse <ns>,  No matching clause: :-`, so this found a bug!

Many of my `s/defn`s have return types of a function call `:- (s/maybe s/Int)` which causes the namespace to fail to parse (as reported) so testing them is a bit trickier

**Comment by borkdude** [comment]
@ccann From the Schema documentation:

```
 - Schema metadata is only processed on top-level arguments.  I.e., you can
   use destructuring, but you must put schema metadata on the top-level
   arguments, not the destructured variables.

   Bad:  (s/defn foo [{:keys [x :- s/Int]}])
   Good: (s/defn foo [{:keys [x]} :- {:x s/Int}])
```
so maybe it's better to not support that in the linter?
The rest of the cases should work with this new binary:

https://1716-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.04.22-alpha-SNAPSHOT-macos-amd64.tar.gz

**Comment by borkdude** [comment]
EDIT: fixed one more bug
https://1719-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.04.22-alpha-SNAPSHOT-macos-amd64.tar.gz

**Comment by ccann** [comment]
Nice! Yeah, I suspected some of these might be a bit niche, or apparently just wrong in once case.

Looks like all those cases are working 👍 👍

**Comment by mynomoto** [comment]
If there is a linux build I would like to test that too :heart_eyes:

**Comment by borkdude** [comment]
@mynomoto https://1724-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.04.22-alpha-SNAPSHOT-linux-amd64.tar.gz

**Comment by mynomoto** [comment]
@borkdude Thanks!

**Comment by borkdude** [comment]
There was one bug where if you typed `(s/defn)` the linter would get into an infinite loop. 
This is fixed in: https://1730-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.04.22-alpha-SNAPSHOT-linux-amd64.tar.gz

**Comment by borkdude** [comment]
Latest Mac version: https://1731-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.04.22-alpha-SNAPSHOT-macos-amd64.tar.gz

If I don't hear anything in the next few hours, I'm going to merge to master and release it :-).

**Comment by ccann** [comment]
Sounds good to me @borkdude 
I don't have time today to do more testing unfortunately, but it looked good on the existing codebases!
Thanks for the awesome tool! 

I'll open a new issue if I discover anything else schema related in the future

**Comment by mynomoto** [comment]
@borkdude on my project everything looks good on s/defn usage. :fireworks:
